### ISSUE TITLE
Push state not maintaining good browser history

### ISSUE THREAD

**Issue by ctrlaltdan** [angular/angular#8707]
**angular 2.0.0-rc.1**

I'm attempting to maintain state by passing parameters to the URL.

It seems when I attempt to do so using the `Location` class the browsers history gets very messy and using the browser back button doesn't cycle back through my various states.

**Steps to reproduce and a minimal demo of the problem**

See http://plnkr.co/edit/5RIIeFlQMds9Tc1t1W0r?p=preview

_What steps should we try in your demo to see the problem?_
- Click show animals
- Click Pages 1 through 5
- Click browser back button a few times

**Current behavior**

URL changes from
- /animals;page=5
- /animals;page=4
- /animals;page=4
- /animals;page=4

**Expected/desired behavior**

URL to change to
- /animals;page=5
- /animals;page=4
- /animals;page=3
- /animals;page=2

**Other information**

Sorry if this is actually by design, given the minimal code required to recreate this I assumed it may be a bug.

Also

http://stackoverflow.com/questions/37278133/how-to-push-state-to-url-in-angular-2-rc-1-and-maintain-good-browser-history

**Comment by dubs999** [comment]
After more investigation it appears that this is what happens

_after navigating from pages 1 through 5_

- Click back button
- URL changes to _last page visited_
- Angular routing then takes over and attempts to interpret the route `/animals;page=n`
- Angular routing decides to route back to `/animals` (after failing to interpret the route?)
- URL remains as `/animals;page=n`
- Subsequent browser back button clicks are actually attempting to go from `/animals` to `/animals;page=n` - this would describe the "Page 5, 4, 4, 4..." symptom.

This isn't quite so simple to repro in plunker as calls to anything other than `index.html` will attempt to navigate to unknown files. This was spotted by downloading the source code as a zip and running locally.

Annoyingly a direct browser call to `/animals;page=3` works fine (when set up in a proper environment). A call to the same URL given the user has clicked the browser back button fails to recognize the route and we get unexpected results.

Again - sorry if this is actually by design.

**Comment by Gamblt** [comment]
I've created several different components with different basic router (like /about, /news, /press) and had the same issue. Browser back history button works only once. :(

**Comment by sod** [comment]
Yes. The Router does a navigate(location.pathname) => location.go => pushState on every popState and with it preventing more than one history back.

https://github.com/angular/angular/blob/master/modules/@angular/router/src/router.ts#L151

**Comment by dloukadakis** [comment]
This probably was fixed by my changes in 595bcdd1ac922fcdc52e50e057d8d85fc0a704d3

**Comment by patrickmichalina** [comment]
This has hit me when using RC5. Can only go back once between the same two routes.

**Comment by adam-s** [comment]
Has this issue been resolved. When I press back it only goes to the previous state. If I press back twice it goes in a circle between the two pages.

**Comment by BenDevelopment** [comment]
I have the same issue as @adam-s . Any workaround ?

**Comment by dloukadakis** [comment]
Can u share some more details? I can't seem to be able to reproduce.

**Comment by BenDevelopment** [comment]
I have a project with submodules.
Each submodule has a modulname-routing.module.ts file.
For example, I have a user-routing.module.ts file which looks like that:
```
const routes: Routes = [
  { path: 'user/:pk', component: ConsultComponent, canActivate: [AuthGuard] },
  { path: 'home', component: HomeComponent, canActivate: [AuthGuard],
    children: [
      { path: '', redirectTo: 'coachs', pathMatch: 'full' },
      { path: 'coachs', component: HomeCoachsComponent },
      { path: 'athletes', component: HomeAthletesComponent },
      { path: 'settings', component: SettingsComponent },
    ]
  },

  { path: 'first-sign-in', component: FirstSignInComponent, canActivate: [AuthGuard] },
];

@NgModule({
  imports: [RouterModule.forRoot(routes)],
  exports: [RouterModule]
})
export class UserRoutingModule { }
```
My user.module.ts file looks like that:
```
@NgModule({
  imports: [
    CommonModule,
    FormsModule,
    UserRoutingModule, // <-- The user routing is imported here
  ],
  declarations: [
  ...
  ],
  providers: [
  ...
  ],
  exports: [
  ...
  ]
})
export class UserModule { }

```
And finally, my app.module which import the user module:
```
@NgModule({
  imports: [
    AppRoutingModule,
    BrowserModule,
    FormsModule,
    HttpModule,
    UserModule, // <- the user module here
    ...
  ],
  declarations: [
    AppComponent,
    ...
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }
```
I have nothing else in my code concerning routing. All works well appart this problem with the back/forward navigation.

**Comment by smilanko** [comment]
bump

**Comment by DzmitryShylovich** [comment]
@dubs999 pls update your plunkr

**Comment by DzmitryShylovich** [comment]
@vicb no repro. can be closed

**Comment by thaod** [comment]
This issue still happens to me with Angular 4.2.4.
After I use `location.go(...)`, neither the back button of browser or `router.navigate()` change browser url.

**Comment by JBeen** [comment]
I've found answer with location.onPopState.
Link: https://stackoverflow.com/questions/40381814/how-do-i-detect-user-navigating-back-in-angular2

**Comment by ajitsinghkaler** [comment]
@gkalpak I think this is resolved.

**Comment by gkalpak** [comment]
Yes, afaict it works as expected now.
If anyone is still experiencing problems, please open a new issue providing an up-to-date reproduction (using the latest Angular version).
